comet-java-client/src/main/java/ml/comet/experiment/impl/BaseExperiment.java

```diff
@@ -0,0 +1,899 @@
+    }
+
+    @Override
+    public List<ValueMinMaxDto> getMetrics() {
+        if (getLogger().isDebugEnabled()) {
+            getLogger().debug("get metrics summary for experiment {}", this.experimentKey);
+        }
+
+        return loadRemote(restApiClient::getMetrics, "METRICS").getValues();
+    }
+
+    @Override
+    public List<ValueMinMaxDto> getLogOther() {
+        if (getLogger().isDebugEnabled()) {
+            getLogger().debug("get log other for experiment {}", this.experimentKey);
+        }
+
+        return loadRemote(restApiClient::getLogOther, "OTHER PARAMETERS").getValues();
+    }
+
+    @Override
+    public List<String> getTags() {
+        if (getLogger().isDebugEnabled()) {
+            getLogger().debug("get tags for experiment {}", this.experimentKey);
+        }
+
+        return loadRemote(restApiClient::getTags, "TAGs").getTags();
+    }
+
+    @Override
+    public List<ExperimentAssetLink> getAssetList(@NonNull AssetType type) {
+        if (getLogger().isDebugEnabled()) {
+            getLogger().debug("get assets with type {} for experiment {}", type, this.experimentKey);
+        }
+
+        return validateAndGetExperimentKey()
+                .concatMap(experimentKey -> restApiClient.getAssetList(experimentKey, type))
+                .doOnError(ex -> getLogger().error("Failed to read ASSETS list for the experiment, experiment key: {}",
+                        this.experimentKey, ex))
+                .blockingGet()
+                .getAssets();
+    }
+
+    @Override
+    public void end() {
+        if (!this.alive) {
+            return;
+        }
+        getLogger().info("Waiting for all scheduled uploads to complete. It can take up to {} seconds.",
+                cleaningTimeout.getSeconds());
+
+        // mark as not alive
+        this.alive = false;
+
+        // close REST API
+        this.restApiClient.dispose();
+
+        // close connection
+        if (this.connection != null) {
+            try {
+                this.connection.waitAndClose(this.cleaningTimeout);
+                this.connection = null;
+            } catch (Exception e) {
+                getLogger().error("failed to close connection", e);
+            }
+        }
+
+        // dispose all pending calls
+        if (disposables.size() > 0) {
+            getLogger().warn("{} calls still has not been processed, disposing", disposables.size());
+        }
+        this.disposables.dispose();
+    }
+
+    /**
+     * Sends heartbeat to the server and returns the status response.
+     *
+     * @return the status response of the experiment.
+     */
+    Optional<ExperimentStatusResponse> sendExperimentStatus() {
+        return Optional.ofNullable(validateAndGetExperimentKey()
+                .concatMap(experimentKey -> restApiClient.sendExperimentStatus(experimentKey))
+                .onErrorComplete()
+                .blockingGet());
```

Comment:
This sends a blocking Get but is run in a background thread so it won't block the main thread execution, is that correct?

Reply:
This method invoked from scheduled thread pool executor to periodically send heartbeats and get experiment status. It will not block the main thread.